tools/internal/ocpaths/ocpaths.go

```diff
@@ -0,0 +1,195 @@
+type OCPathKey struct {
+	Path      string
+	Component string
+}
+
+// OCPath is the parsed version of the spreadsheet's paths.
+type OCPath struct {
+	Key              OCPathKey
+	FeatureprofileID string
+}
+
+func getSchemaFakeroot(publicPath string) (*yang.Entry, error) {
+	files, err := yangutil.GetAllYANGFiles(publicPath)
+	if err != nil {
+		return nil, err
+	}
+
+	moduleEntryMap, errs := yangentry.Parse(files, []string{publicPath})
+	if errs != nil {
+		return nil, err
+	}
+	root := &yang.Entry{
+		Dir: map[string]*yang.Entry{},
+	}
+	for _, entry := range moduleEntryMap {
+		// Skip IETF modules.
+		if !strings.HasPrefix(entry.Name, "openconfig-") {
+			continue
+		}
+		for name, ch := range entry.Dir {
+			root.Dir[name] = ch
+		}
+	}
+
+	return root, nil
+}
+
+func validatePath(ocpathProto *ppb.OCPath, root *yang.Entry) (*OCPath, error) {
+	ocpath := &OCPath{
+		Key: OCPathKey{
+			Path:      ocpathProto.GetName(),
+			Component: ocpathProto.GetOcpathConstraint().GetPlatformType(),
+		},
+		FeatureprofileID: ocpathProto.GetFeatureprofileid(),
+	}
+
+	// Validate path
+	path := ocpath.Key.Path
+	if !strings.HasPrefix(path, "/") {
+		return nil, fmt.Errorf("path does not begin with slash: %q", path)
+	}
+	if strings.HasSuffix(path, "/") {
+		return nil, fmt.Errorf("path must not end with slash: %q", path)
+	}
+	if entry := root.Find(path); entry == nil {
+		deepestEntry := root
+		var entryNotFound string
+		splitEles := strings.Split(path, "/")
+		for i, ele := range splitEles {
+			// Skip the first element, which must be empty.
+			if i == 0 {
+				continue
+			}
+			next := deepestEntry.Dir[ele]
+			if next == nil {
+				entryNotFound = strings.Join(splitEles[i:], "/")
+				break
+			}
+			deepestEntry = next
+		}
+		return nil, fmt.Errorf("path not found: %q, remaining path: %q", path, entryNotFound)
+	} else if !entry.IsLeaf() && !entry.IsLeafList() {
+		return nil, fmt.Errorf("path %q is not a leaf: got kind %s", path, entry.Kind)
+	}
+
+	// Validate component
+	component := ocpath.Key.Component
+	isComponent := strings.HasPrefix(path, componentPrefix)
+	matched, err := regexp.MatchString("^[A-Z_]+$", component)
+	if err != nil {
+		return nil, err
+	}
+CONSTRAINT_CHECK:
+	switch {
+	case !isComponent && component != "":
+		return nil, fmt.Errorf("non-component path %q has component value %q", path, component)
+	case !isComponent:
+		// valid
+	case matched:
+		for _, enum := range []string{
+			reflect.TypeOf(oc.E_PlatformTypes_OPENCONFIG_HARDWARE_COMPONENT(0)).Name(),
+			reflect.TypeOf(oc.E_PlatformTypes_OPENCONFIG_SOFTWARE_COMPONENT(0)).Name(),
+		} {
+			for _, v := range oc.ΛEnum[enum] {
+				if v.Name == component {
+					ocpath.Key.Component = component
+					break CONSTRAINT_CHECK
+				}
+			}
+		}
+		fallthrough
+	default:
+		return nil, fmt.Errorf("path %q has invalid component %q", path, component)
+	}
+
+	// Basic validation of featureprofileid if it exists
```

Comment:
Be even more explicit.  featureprofileID is optional.  Only validate the string format if it exists.

Reply:
Done.